## Nama : Syerli Khuntum Khaira
## NIM : 4222301028
## Kelas : RE 6A Pagi
## Practice 7&8


In [27]:
#import library
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

In [28]:
#memuat dataset
df = pd.read_csv('iris.data.csv', header=None)
print("Dataset iris berhasil\n")

df.columns = [
    'sepal_length',
    'sepal_width',
    'petal_length',
    'petal_width',
    'class'
]
print("Ukuran Dataset :", df.shape)

df.head()

Dataset iris berhasil

Ukuran Dataset : (150, 5)


,sepal_length,sepal_width,petal_length,petal_width,class
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


In [29]:
df.info()
df.describe()
df["class"].value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   class         150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


,count
class,
Iris-setosa,50
Iris-versicolor,50
Iris-virginica,50


In [30]:
# feature dan target
X = df.drop("class", axis=1)

y = df["class"]

print("Shape X :", X.shape)
print("Shape y :", y.shape)


Shape X : (150, 4)
Shape y : (150,)


In [31]:
# Encode label
encoder = LabelEncoder()
y = encoder.fit_transform(y)

print("Label Iris:")
for i, kelas in enumerate(encoder.classes_):
    print(f"{i} = {kelas}")


Label Iris:
0 = Iris-setosa
1 = Iris-versicolor
2 = Iris-virginica


In [32]:
# split dataset
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.3,
    random_state=42,
    stratify=y
)

print("Data Train :", X_train.shape)
print("Data Test  :", X_test.shape)

Data Train : (105, 4)
Data Test  : (45, 4)


In [33]:
# fitur standardisasi scaler
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [34]:
# fungsi aktivasi
activations = {
    "logistic": "Sigmoid",
    "tanh": "Tanh",
    "relu": "Relu"
}


In [35]:
#mempresentasi fungsi aktivasi
results = []

print("Fungsi Aktivasi Dijalankan...")

for act_code, act_name in activations.items():

    mlp = MLPClassifier(
        hidden_layer_sizes=(10,10),
        activation=act_code,
        solver='adam',
        learning_rate_init=0.001,
        max_iter=1000,
        random_state=42
    )

    mlp.fit(X_train, y_train)

    y_pred = mlp.predict(X_test)

    acc = accuracy_score(y_test, y_pred)
    prec = precision_score(y_test, y_pred, average='weighted')
    rec = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')

    results.append([
        act_name,
        acc,
        prec,
        rec,
        f1
    ])

Fungsi Aktivasi Dijalankan...


/usr/local/lib/python3.12/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


In [38]:
results_formatted = results_df.copy()

for col in ['Accuracy', 'Precision (weighted)', 'Recall (weighted)', 'F1-Score (weighted)']:
    results_formatted[col] = results_formatted[col].apply(
        lambda x: f"{x*100:.2f}%"
    )

print("\n--- TABEL PERBANDINGAN PERFORMA MLP ---")
print(results_formatted.to_string(index=False))


--- TABEL PERBANDINGAN PERFORMA MLP ---
Activation Function Accuracy Precision (weighted) Recall (weighted) F1-Score (weighted)
            Sigmoid   93.33%               93.45%            93.33%              93.33%
               Tanh   95.56%               95.56%            95.56%              95.56%
               Relu   91.11%               91.55%            91.11%              91.07%


In [37]:
results_df = pd.DataFrame(
    results,
    columns=[
        'Activation Function',
        'Accuracy',
        'Precision (weighted)',
        'Recall (weighted)',
        'F1-Score (weighted)'
    ]
)

results_formatted = results_df.copy()

for col in [
    'Accuracy',
    'Precision (weighted)',
    'Recall (weighted)',
    'F1-Score (weighted)'
]:
    results_formatted[col] = results_formatted[col].apply(
        lambda x: f"{x*100:.2f}%"
    )

print("Tabel Perbandingan Performa Fungsi Aktivasi pada MLP:")
display(results_formatted)

Tabel Perbandingan Performa Fungsi Aktivasi pada MLP:


,Activation Function,Accuracy,Precision (weighted),Recall (weighted),F1-Score (weighted)
0,Sigmoid,93.33%,93.45%,93.33%,93.33%
1,Tanh,95.56%,95.56%,95.56%,95.56%
2,Relu,91.11%,91.55%,91.11%,91.07%


## Catatan Hasil Perbandingan Fungsi Aktivasi pada MLP

Berdasarkan hasil pengujian menggunakan dataset Iris, fungsi aktivasi **Tanh**
memberikan performa terbaik dibandingkan Sigmoid dan ReLU. Hal ini ditunjukkan
oleh nilai Accuracy, Precision, Recall, dan F1-Score yang paling tinggi, yaitu

sebesar **95.56%**.
Sementara itu, fungsi aktivasi Sigmoid menghasilkan Accuracy sebesar **93.33%**, sedangkan ReLU menghasilkan Accuracy sebesar **91.11%**. Meskipun ketiga fungsi aktivasi mampu melakukan klasifikasi dengan baik, Tanh memberikan hasil yang lebih konsisten pada seluruh metrik evaluasi.

Oleh karena itu, dapat disimpulkan bahwa **fungsi aktivasi Tanh merupakan fungsi aktivasi terbaik untuk dataset Iris pada percobaan ini**, karena menghasilkan performa klasifikasi tertinggi dibandingkan fungsi aktivasi Sigmoid dan ReLU.
